// x12-parser/dist/esm/X12grouper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.X12grouper = void 0;
const node_stream_1 = require("node:stream");
const Schema_js_1 = require("./Schema.js");
const Group_js_1 = require("./Group.js");
class X12grouper extends node_stream_1.Transform {
    #schemas;
    #version;
    #activeGroup;
    #defaultSchema;
    #initialHold;
    /**
     * Creates a new Validator stream
     * @param {Schema[] | Schema} schemas A Schema or array of Schemas the grouper can use to dynamically select based on X12 file version
     */
    constructor(schemas) {
        super({ objectMode: true });
        this.#schemas = Array.isArray(schemas) ? schemas : [schemas];
        this.#version = '';
        this.#activeGroup = null;
        this.#defaultSchema =
            this.#schemas.find((schema) => schema.default) || this.#schemas[0];
        this.#initialHold = [];
        this.#schemas.forEach((schema) => {
            if (!(schema instanceof Schema_js_1.Schema)) {
                throw new TypeError('Every schema must be a Schema object');
            }
        });
    }
    get schemas() {
        return this.#schemas;
    }
    get activeVersion() {
        return this.#version;
    }
    get activeGroup() {
        return this.#activeGroup;
    }
    get initialHold() {
        return this.#initialHold;
    }
    get defaultSchema() {
        return this.#defaultSchema;
    }
    /**
     * Do not directly use, this is called by readable stream methods when stream has finished
     * @param cb Execute cb when done flushing
     */
    _flush(cb) {
        if (this.#activeGroup) {
            this.groupDone(this.#activeGroup);
        }
        cb();
    }
    /**
     * Node API for transform streams
     * @param chunk The next chunk of data from read stream
     * @param encoding The encoding of the chunk (e.g. UTF-8)
     * @param cb Callback indicating chunk was processed and ready for next chunk
     */
    _transform(chunk, _encoding, cb) {
        this.processSegment(chunk);
        cb();
    }
    //TODO: Add validation to GS
    set version(GS) {
        this.#version = GS['8'];
        this.#initialHold.forEach((segment) => this.appendToGroup(segment));
        this.#initialHold = [];
    }
    newGroup(segment) {
        const match = this.#schemas.find((item) => item.version === this.#version) ||
            this.#defaultSchema;
        // If there is no active groups and it does not match the segment name used to start the group emit this segment without grouping
        if (segment.name !== match.schema.start && !this.#activeGroup) {
            this.push(segment);
        }
        else {
            this.#activeGroup = new Group_js_1.Group(match.schema, segment, this.groupDone.bind(this));
        }
    }
    /**
     * Sends the grouped data down the stream and resets active group
     */
    groupDone(group) {
        this.push(group.data);
        this.#activeGroup = null;
    }
    /**
     * Adds data to the active group or starts a new group
     */
    appendToGroup(segment) {
        if (!this.#activeGroup) {
            this.newGroup(segment);
        }
        else if (!this.#activeGroup.add(segment)) {
            this.newGroup(segment);
        }
    }
    processSegment(segment) {
        // Since we select version from GS we need to hold ISA in case it's part of schema
        if (segment.name === 'ISA') {
            this.#initialHold.push(segment);
            return;
        }
        else if (segment.name === 'GS') {
            this.version = segment;
        }
        this.appendToGroup(segment);
    }
}
exports.X12grouper = X12grouper;
